File: 10154 Weights and Measures/solution.java
// #解法一
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
        ArrayList<Turtle> turtles = new ArrayList<>();
        StreamTokenizer st = new StreamTokenizer(System.in);
        while(st.nextToken() == StreamTokenizer.TT_NUMBER) {
            int w = (int)st.nval;
            st.nextToken();
            turtles.add(new Turtle(w, (int)st.nval));
        }
        turtles.sort(new Comparator<Turtle>() {
            @Override
            public int compare(Turtle t1, Turtle t2) {
                return Integer.compare(t1.strength, t2.strength);
            }
        });
        ArrayList<Integer> stacked = new ArrayList<>();
        int totalWeight = 0, height = 0;
        for(Turtle t : turtles) {
            if(totalWeight + t.weight <= t.strength) {
                stacked.add(t.weight);
                totalWeight += t.weight;
                ++height;
            } else if(height > 0) {
                int maxIndex = 0;
                for(int i = 1; i < height; ++i) {
                    if(stacked.get(i) > stacked.get(maxIndex))
                        maxIndex = i;
                }
                if(stacked.get(maxIndex) > t.weight) {
                    totalWeight += t.weight - stacked.get(maxIndex);
                    stacked.set(maxIndex, t.weight);
                }
            }
        }
	    System.out.println(height);
	}
}

class Turtle {
    int weight, strength;
    Turtle(int weight, int strength) {
        this.weight = weight;
        this.strength = strength;
    }
}

// #解法二
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
        ArrayList<Turtle> turtles = new ArrayList<>();
        StreamTokenizer st = new StreamTokenizer(System.in);
        while(st.nextToken() == StreamTokenizer.TT_NUMBER) {
            int w = (int)st.nval;
            st.nextToken();
            turtles.add(new Turtle(w, (int)st.nval));
        }
        turtles.sort(new Comparator<Turtle>() {
            @Override
            public int compare(Turtle t1, Turtle t2) {
                return Integer.compare(t1.strength, t2.strength);
            }
        });
        int[] dp = new int[turtles.size() + 1];
        Arrays.fill(dp, 1, turtles.size(), 1000000000);
        int maxHeight = 0;
        for(Turtle t : turtles) {
            for(int i = maxHeight; i >= 0; --i) {
                if(dp[i] + t.weight <= t.strength) {
                    dp[i + 1] = Math.min(dp[i + 1], dp[i] + t.weight);
                    maxHeight = Math.max(maxHeight, i + 1);
                }
            }
        }
	    System.out.println(maxHeight);
	}
}

class Turtle {
    int weight, strength;
    Turtle(int weight, int strength) {
        this.weight = weight;
        this.strength = strength;
    }
}

File: 10154 Weights and Measures/solution.c
// #解法一
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int weight, strength;
} Turtle;

int compare(const void *a, const void *b) {
    return ((Turtle*)a)->strength > ((Turtle*)b)->strength;
}

int main() {
    Turtle *turtles = NULL;
    int size = 0, w, s;
    while(scanf("%d %d", &w, &s) == 2) {
        turtles = (Turtle*)realloc(turtles, (size + 1) * sizeof(Turtle));
        turtles[size++] = (Turtle){w, s};
    }
    qsort(turtles, size, sizeof(Turtle), compare);
    int *stacked = NULL, totalWeight = 0, height = 0;
    for(int i = 0; i < size; ++i) {
        if(totalWeight + turtles[i].weight <= turtles[i].strength) {
            stacked = (int*)realloc(stacked, (height + 1) * sizeof(int));
            totalWeight += stacked[height++] = turtles[i].weight;
        } else if(height > 0) {
            int maxIndex = 0;
            for(int j = 1; j < height; ++j) {
                if(stacked[j] > stacked[maxIndex])
                    maxIndex = j;
            }
            if(stacked[maxIndex] > turtles[i].weight) {
                totalWeight += turtles[i].weight - stacked[maxIndex];
                stacked[maxIndex] = turtles[i].weight;
            }
        }
    }
    printf("%d\n", height);
    free(stacked);
    free(turtles);
	return 0;
}

// #解法二
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int weight, strength;
} Turtle;

int compare(const void *a, const void *b) {
    return ((Turtle*)a)->strength > ((Turtle*)b)->strength;
}

int main() {
    Turtle *turtles = NULL;
    int size = 0, w, s;
    while(scanf("%d %d", &w, &s) == 2) {
        turtles = (Turtle*)realloc(turtles, (size + 1) * sizeof(Turtle));
        turtles[size++] = (Turtle){w, s};
    }
    qsort(turtles, size, sizeof(Turtle), compare);
    int dp[size + 1], maxHeight = 0;
    dp[0] = 0;
    for(int i = 0; i < size; ++i) {
        dp[i + 1] = 1e9;
        for(int j = maxHeight; j >= 0; --j) {
            if(dp[j] + turtles[i].weight <= turtles[i].strength) {
                dp[j + 1] = dp[j + 1] < dp[j] + turtles[i].weight ? dp[j + 1] : dp[j] + turtles[i].weight;
                maxHeight = maxHeight > j + 1 ? maxHeight : j + 1;
            }
        }
    }
    printf("%d\n", maxHeight);
    free(turtles);
	return 0;
}
